File: Development/GuitarTeacher/src/com/mad/guitarteacher/chordGeneration/NoteFingerPosition.java
package com.mad.guitarteacher.chordGeneration;

import com.mad.lib.musicalBase.Notes;
import com.mad.tunerlib.chordGeneration.FingerPosition;

public class NoteFingerPosition
{
	public FingerPosition	fingerPosition;
	public int				Note;

	public NoteFingerPosition(FingerPosition cPosition, int eNote)
	{
		fingerPosition = new FingerPosition(cPosition);
		Note = eNote;
	}

	public NoteFingerPosition(NoteFingerPosition other)
	{
		fingerPosition =
				new FingerPosition(other.fingerPosition);
		Note = other.Note;
	}

	public boolean advanceFrets(int frets)
	{
		if (!fingerPosition.advanceFrets(frets))
		{
			return false;
		}

		if (frets > 0)
		{
			if ((fingerPosition.Fret + FingerPosition
					.sameNoteFretDifferenceOnString(
							fingerPosition.String,
							fingerPosition.String + 1)) >= 0)
			{
				if (!fingerPosition.NextString())
				{
					return fingerPosition.PreviousOctave();
				}
			}
		}
		else
		{
			if ((fingerPosition.Fret % FingerPosition
					.sameNoteFretDifferenceOnString(
							fingerPosition.String,
							fingerPosition.String - 1)) == 0)
			{
				// Can fail.
				fingerPosition.previousString();
			}
		}

		int nCurrNote = Note;
		int nNumOfNotes = Notes.NumberOfNotes;

		// Advance notes
		Note =
				((nNumOfNotes + ((nCurrNote + frets) % nNumOfNotes)) % nNumOfNotes);

		return true;
	}

	public boolean nextString()
	{
		return fingerPosition.NextString();
	}

	public boolean previousString()
	{
		return fingerPosition.previousString();
	}
};

File: Development/ChordFingerFinder/NoteFingerPosition.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ChordFingerFinder
{
    class CNoteFingerPosition
    {
        public CFingerPosition FingerPosition { get; private set; }
        public int Note { get; private set; }

        public CNoteFingerPosition(CFingerPosition cPosition, int eNote)
        {
            FingerPosition = new CFingerPosition(cPosition);
            Note = eNote;
        }

        public CNoteFingerPosition(CNoteFingerPosition other)
        {
            FingerPosition = new CFingerPosition(other.FingerPosition);
            Note = other.Note;
        }

        public bool AdvanceFrets(int nFrets)
        {
            if (!FingerPosition.AdvanceFrets(nFrets))
            {
                return false;
            }

            if(nFrets > 0)
            {
                if (FingerPosition.Fret + FingerPosition.SameNoteFretDifferenceOnString(FingerPosition.String, FingerPosition.String+1) >= 0)
                {
                    if (!FingerPosition.NextString())
                    {
                        return FingerPosition.PreviousOctave();
                    }
                }
            }
            else
            {
                if (FingerPosition.Fret % FingerPosition.SameNoteFretDifferenceOnString(FingerPosition.String, FingerPosition.String - 1) == 0)
                {
                    // Can fail.
                    FingerPosition.PreviousString();
                }
            }

            int nCurrNote = (int)Note;
            int nNumOfNotes = (int)ENotes.NumberOfNotes;

            // Advance notes
            Note = ((nNumOfNotes + ((nCurrNote + nFrets) % nNumOfNotes)) % nNumOfNotes);

            return true;
        }

        public bool NextString()
        {
            return FingerPosition.NextString();
        }

        public bool PreviousString()
        {
            return FingerPosition.PreviousString();
        }
    };
}
